crates/ruff_python_formatter/src/comments/format.rs

```diff
@@ -373,3 +399,44 @@ impl Format<PyFormatContext<'_>> for FormatEmptyLines {
         }
     }
 }
+
+// TODO(cnpryer): Alternatively could just duplicate the range tracking from normalization instead
+//   of costly String operations and allocations.
+//   Could abstract the comment behind `NormalizedComment`
+fn normalize_comment<'a>(comment: &'a SourceComment, source: SourceCode<'a>) -> Cow<'a, str> {
+    let slice = comment.slice();
+    let comment_text = slice.text(source);
+
+    // If the comment isn't leading with '#' it's invalid, so we return the text as-is borrowed.
+    // TODO(cnpryer): If this is always used alongside `.fmt()` on comments this should be safe, but
+    //   we can probably do better here.
+    let Some(content) = comment_text.strip_prefix('#') else {
+        return Cow::Borrowed(comment_text);
+    };
+
+    // Make the content mutable in case we need to normalize it. Any normalization is done with lazy operations.
+    let mut content = Cow::Borrowed(content);
+
+    // If the comment has trailing whitespace then we take ownership of a string to make mutations to
+    let trimmed = content.trim_end();
+    let trailing_whitespace_len = content.text_len() - trimmed.text_len();
+    if trailing_whitespace_len > TextSize::new(0) {
+        content = Cow::Owned(trimmed.to_owned());
+    }
+
+    // Fast path for correctly formatted comments:
+    // * Start with a `# '.
+    // * Have no trailing whitespace.
+    if trailing_whitespace_len == TextSize::new(0) && content.starts_with(' ') {
+        return Cow::Borrowed(comment_text);
+    }
+
+    // Formatted comments start with '# '. We perform this normalization if it's necessary.
+    if !content.is_empty() && !content.starts_with([' ', '!', ':', '#', '\'', '\u{A0}']) {
```

Comment:
So far `normalize_comment` is basically a duplication of the normalization from `FormatComment`'s `fmt` implementation. I also need to make sure to handle this comment if I'm not already doing so
https://github.com/astral-sh/ruff/blob/fc89976c24d1c8b9d13d88ef70b619e8a2ac393a/crates/ruff_python_formatter/src/comments/format.rs#L317-L318

Reply:
See https://github.com/astral-sh/ruff/pull/6901#discussion_r1306714876

Reply:
Yes, but we could change `FormatComment` to use `normalize_comment` internally to avoid the duplication.

Reply:
Resolving this to move to https://github.com/astral-sh/ruff/pull/6901/files#r1307947658